pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go

```diff
@@ -0,0 +1,351 @@
+	policyRunner := resiliency.NewRunner[jwk.Key](ctx,
+		a.Resiliency.ComponentOutboundPolicy(in.ComponentName, resiliency.Crypto),
+	)
+	start := time.Now()
+	res, err := policyRunner(func(ctx context.Context) (jwk.Key, error) {
+		return component.GetKey(ctx, in.Name)
+	})
+	elapsed := diag.ElapsedSince(start)
+
+	diag.DefaultComponentMonitoring.CryptoInvoked(ctx, in.ComponentName, diag.Get, err == nil, elapsed)
+
+	if err != nil {
+		err = messages.ErrCryptoGetKey.WithFormat(in.Name, err)
+		a.Logger.Debug(err)
+		return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+	}
+
+	// Get the key ID if present
+	kid := in.Name
+	if dk, ok := res.(*contribCrypto.Key); ok {
+		kid = dk.KeyID()
+	}
+
+	// Format the response
+	var pk []byte
+	switch in.Format {
+	case runtimev1pb.SubtleGetKeyAlpha1Request_PEM: //nolint:nosnakecase
+		var (
+			v   crypto.PublicKey
+			der []byte
+		)
+		err = res.Raw(&v)
+		if err != nil {
+			err = fmt.Errorf("failed to marshal public key %s as PKIX: %w", in.Name, err)
+			err = messages.ErrCryptoGetKey.WithFormat(in.Name, err)
+			a.Logger.Debug(err)
+			return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+		}
+		der, err = x509.MarshalPKIXPublicKey(v)
+		if err != nil {
+			err = fmt.Errorf("failed to marshal public key %s as PKIX: %w", in.Name, err)
+			err = messages.ErrCryptoGetKey.WithFormat(in.Name, err)
+			a.Logger.Debug(err)
+			return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+		}
+		pk = pem.EncodeToMemory(&pem.Block{
+			Type:  "PUBLIC KEY",
+			Bytes: der,
+		})
+
+	case runtimev1pb.SubtleGetKeyAlpha1Request_JSON: //nolint:nosnakecase
+		pk, err = json.Marshal(res)
+		if err != nil {
+			err = fmt.Errorf("failed to marshal public key %s as JSON: %w", in.Name, err)
+			err = messages.ErrCryptoGetKey.WithFormat(in.Name, err)
+			a.Logger.Debug(err)
+			return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+		}
+
+	default:
+		err = messages.ErrBadRequest.WithFormat("invalid key format")
+		a.Logger.Debug(err)
+		return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+	}
+
+	return &runtimev1pb.SubtleGetKeyAlpha1Response{
+		Name:      kid,
+		PublicKey: string(pk),
+	}, nil
+}
+
+type subtleEncryptRes struct {
+	ciphertext []byte
+	tag        []byte
+}
+
+// SubtleEncryptAlpha1 encrypts a small message using a key stored in the vault.
+func (a *UniversalAPI) SubtleEncryptAlpha1(ctx context.Context, in *runtimev1pb.SubtleEncryptAlpha1Request) (*runtimev1pb.SubtleEncryptAlpha1Response, error) {
+	component, err := a.cryptoValidateRequest(in.ComponentName)
+	if err != nil {
+		return &runtimev1pb.SubtleEncryptAlpha1Response{}, err
+	}
+
+	policyRunner := resiliency.NewRunner[subtleEncryptRes](ctx,
+		a.Resiliency.ComponentOutboundPolicy(in.ComponentName, resiliency.Crypto),
+	)
+	start := time.Now()
+	ser, err := policyRunner(func(ctx context.Context) (r subtleEncryptRes, rErr error) {
+		r.ciphertext, r.tag, rErr = component.Encrypt(ctx, in.Plaintext, in.Algorithm, in.Key, in.Nonce, in.AssociatedData)
```

Comment:
can the algorithm be validated before calling the component?

Reply:
Are you thinking of using the `SupportedEncryptionAlgorithms` method?

Reply:
Yes

Reply:
IMHO that's redundant. All components right now synchronously check if the algorithm is supported at the very beginning. In the case of AKV, it's checked before invoking the cloud service: https://github.com/dapr/components-contrib/blob/master/crypto/azure/keyvault/component.go#L147

I'd let the component make that decision personally. There can also be edge cases the component can handle better than the runtime.